EKF/airspeed_fusion.cpp

```diff
@@ -46,152 +47,137 @@
-		Vector24f H_TAS;
-		H_TAS(4) = SH_TAS[2];
-		H_TAS(5) = SH_TAS[1];
-		H_TAS(6) = vd*SH_TAS[0];
-		H_TAS(22) = -SH_TAS[2];
-		H_TAS(23) = -SH_TAS[1];
-
-		_airspeed_innov_var = (R_TAS + SH_TAS[2]*(P(4,4)*SH_TAS[2] + P(5,4)*SH_TAS[1] - P(22,4)*SH_TAS[2] - P(23,4)*SH_TAS[1] + P(6,4)*vd*SH_TAS[0]) + SH_TAS[1]*(P(4,5)*SH_TAS[2] + P(5,5)*SH_TAS[1] - P(22,5)*SH_TAS[2] - P(23,5)*SH_TAS[1] + P(6,5)*vd*SH_TAS[0]) - SH_TAS[2]*(P(4,22)*SH_TAS[2] + P(5,22)*SH_TAS[1] - P(22,22)*SH_TAS[2] - P(23,22)*SH_TAS[1] + P(6,22)*vd*SH_TAS[0]) - SH_TAS[1]*(P(4,23)*SH_TAS[2] + P(5,23)*SH_TAS[1] - P(22,23)*SH_TAS[2] - P(23,23)*SH_TAS[1] + P(6,23)*vd*SH_TAS[0]) + vd*SH_TAS[0]*(P(4,6)*SH_TAS[2] + P(5,6)*SH_TAS[1] - P(22,6)*SH_TAS[2] - P(23,6)*SH_TAS[1] + P(6,6)*vd*SH_TAS[0]));
-
-		if (_airspeed_innov_var >= R_TAS) { // Check for badly conditioned calculation
-			_fault_status.flags.bad_airspeed = false;
-
-		} else { // Reset the estimator covariance matrix
-			_fault_status.flags.bad_airspeed = true;
-
-			// if we are getting aiding from other sources, warn and reset the wind states and covariances only
-			const char* action_string = nullptr;
-			if (update_wind_only) {
-				resetWindStates();
-				resetWindCovariance();
-				action_string = "wind";
-
-			} else {
-				initialiseCovariance();
-				_state.wind_vel.setZero();
-				action_string = "full";
-			}
-			ECL_ERR("airspeed badly conditioned - %s covariance reset", action_string);
-
-			return;
-		}
+	// determine if we need the sideslip fusion to correct states other than wind
+	const bool update_wind_only = !_is_wind_dead_reckoning;
+
+	// Intermediate variables
+	const float HK0 = vn - vwn;
+	const float HK1 = ve - vwe;
+	const float HK2 = ecl::powf(HK0, 2) + ecl::powf(HK1, 2) + ecl::powf(vd, 2);
+	if (HK2 < 1.0f) {
+		// calculation can be badly conditioned for very low airspeed values so don't fuse this time
+		return;
+	}
+	const float v_tas_pred = sqrtf(HK2); // predicted airspeed
+	//const float HK3 = powf(HK2, -1.0F/2.0F);
+	const float HK3 = 1.0f / v_tas_pred;
+	const float HK4 = HK0*HK3;
+	const float HK5 = HK1*HK3;
+	const float HK6 = 1.0F/HK2;
+	const float HK7 = HK0*P(4,6) - HK0*P(6,22) + HK1*P(5,6) - HK1*P(6,23) + P(6,6)*vd;
+	const float HK8 = HK1*P(5,23);
+	const float HK9 = HK0*P(4,5) - HK0*P(5,22) + HK1*P(5,5) - HK8 + P(5,6)*vd;
+	const float HK10 = HK1*HK6;
+	const float HK11 = HK0*P(4,22);
+	const float HK12 = HK0*P(4,4) - HK1*P(4,23) + HK1*P(4,5) - HK11 + P(4,6)*vd;
+	const float HK13 = HK0*HK6;
+	const float HK14 = -HK0*P(22,23) + HK0*P(4,23) - HK1*P(23,23) + HK8 + P(6,23)*vd;
+	const float HK15 = -HK0*P(22,22) - HK1*P(22,23) + HK1*P(5,22) + HK11 + P(6,22)*vd;
+	//const float HK16 = HK3/(-HK10*HK14 + HK10*HK9 + HK12*HK13 - HK13*HK15 + HK6*HK7*vd + R_TAS);
+	float HK16;
+
+	// innovation variance
+	_airspeed_innov_var = (-HK10*HK14 + HK10*HK9 + HK12*HK13 - HK13*HK15 + HK6*HK7*vd + R_TAS);
+	if (_airspeed_innov_var >= R_TAS) { // Check for badly conditioned calculation
```

Comment:
It is a bit cleaner if there is only a single `if (_airspeed_innov_var < R_TAS)` case which returns.